chain/db/statedb.go

```diff
@@ -2,14 +2,279 @@ package db
+	if err != nil {
+		return fmt.Errorf("Deserialize nanoPayCleanup error: %v", err)
+	}
+
+	nanoPayCleanup := make(map[uint32]map[string]struct{}, npcsLength)
+	for i := uint64(0); i < npcsLength; i++ {
+		k, err := serialization.ReadUint32(r)
+		if err != nil {
+			return fmt.Errorf("Deserialize nanoPayCleanup error: %v", err)
+		}
+
+		npcLength, err := serialization.ReadVarUint(r, 0)
+		if err != nil {
+			return fmt.Errorf("Deserialize nanoPayCleanup error: %v", err)
+		}
+
+		v := make(map[string]struct{}, npcLength)
+
+		for j := uint64(0); j < npcLength; j++ {
+			id, err := serialization.ReadVarString(r)
+			if err != nil {
+				return fmt.Errorf("Deserialize nanoPayCleanup error: %v", err)
+			}
+			v[id] = struct{}{}
+		}
+		nanoPayCleanup[k] = v
+	}
+	ga.nanoPayCleanup = nanoPayCleanup
+
+	return nil
+}
+
+func (ga *globalAccount) Empty() bool {
+	return len(ga.nanoPay) == 0 && len(ga.nanoPayCleanup) == 0
+}
+
+func getNanoPayId(sender, recipient common.Uint160, nonce uint64) string {
+	buf := new(bytes.Buffer)
+	_, _ = sender.Serialize(buf)
+	_, _ = recipient.Serialize(buf)
+	_ = serialization.WriteUint64(buf, nonce)
+	hash := sha256.New()
+	hash.Write(buf.Bytes())
+	return string(hash.Sum(nil))
+}
+
+func (ga *globalAccount) GetNanoPay(sender, recipient common.Uint160, nonce uint64) (common.Fixed64, uint32) {
+	id := getNanoPayId(sender, recipient, nonce)
+	if np, ok := ga.nanoPay[id]; ok {
+		return np.balance, np.limit
+	}
+	return 0, 0
+}
+
+func (sdb *StateDB) GetNanoPay(sender, recipient common.Uint160, nonce uint64) (common.Fixed64, uint32) {
+	acc, err := sdb.getAccount(globalAddr, newGlobalAccount)
+	if err != nil {
+		return 0, 0
+	}
+	return acc.(*globalAccount).GetNanoPay(sender, recipient, nonce)
+}
+
+func (ga *globalAccount) cleanupNanoPayAtHeight(height uint32, id string) {
+	ids, ok := ga.nanoPayCleanup[height]
+	if !ok {
+		ids = make(map[string]struct{})
+		ga.nanoPayCleanup[height] = ids
+	}
+	ids[id] = struct{}{}
+}
+
+func (ga *globalAccount) cancelNanoPayCleanupAtHeight(height uint32, id string) {
+	if ids, ok := ga.nanoPayCleanup[height]; ok {
+		if _, ok = ids[id]; ok {
+			delete(ids, id)
+			if len(ids) == 0 {
+				delete(ga.nanoPayCleanup, height)
+			}
+		}
+	}
+}
+
+func (ga *globalAccount) updateNanoPayCleanup(oldHeight, newHeight uint32, id string) {
+	ga.cancelNanoPayCleanupAtHeight(oldHeight, id)
+	ga.cleanupNanoPayAtHeight(newHeight, id)
+}
+
+func (ga *globalAccount) CleanupNanoPay(height uint32) {
+	if ids, ok := ga.nanoPayCleanup[height]; ok {
+		for id := range ids {
+			delete(ga.nanoPay, id)
+		}
+		delete(ga.nanoPayCleanup, height)
+	}
+}
+
+func (ga *globalAccount) SetNanoPay(sender, recipient common.Uint160, nonce uint64, balance common.Fixed64, limit uint32) {
+	id := getNanoPayId(sender, recipient, nonce)
+	var np *nanoPay
+	var ok bool
+	if np, ok = ga.nanoPay[id]; !ok {
+		np = &nanoPay{balance: balance, limit: limit}
+		ga.nanoPay[id] = np
```

Comment:
Should we call cleanupNanoPayAtHeight() here?

Reply:
done